File: Leetcode_with_Cpp/70climbStairs.cpp
class Solution {
public:
    int climbStairs(int n) {
        if (n <= 2) return n;
        int pre2 = 1, pre1 = 2, cur;
        for (int i = 2; i < n; ++i) {
            cur = pre2 + pre1;
            pre2 = pre1;
            pre1 = cur;
        }
        return cur;
    }
};

class Solution {
public:
    int climbStairs(int n) {
        if (n <= 2) return n;
        vector<int> dp(n+1, 1);
        for (int i = 2; i <= n; ++i) 
            dp[i] = dp[i-1] + dp[i-2];
        return dp[n];
    }
};

File: py/70climbStairs.py
class Solution(object):
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        i1 = 1
        i2 = 2
        if n == 1:
            return i1
        for i in range(3, n+1):
            temp = i1 + i2
            i1 = i2
            i2 = temp
        return i2

# dp[i] = dp[i-1] + dp[i-2]:
# 第i次上楼梯方法数 = 第i-1次上楼梯方法数（再走一步） + 第i-2次上楼梯方法数（再走两步）